File: plldb-server/src/main/java/cn/twincest/plldb/util/ContextUtil.java
package cn.twincest.plldb.util;

import cn.twincest.plldb.dao.UserDao;
import cn.twincest.plldb.domain.User;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

public class ContextUtil {

	// session attribute setter and getter

	private static void set(HttpServletRequest request, User user, String rawPassword) {
		request.getSession().setAttribute("user", user);
		request.getSession().setAttribute("rawPassword", rawPassword);
	}

	private static User getUser(HttpServletRequest request) {
		return (User) request.getSession().getAttribute("user");
	}

	private static String getRawPassword(HttpServletRequest request) {
		return (String) request.getSession().getAttribute("rawPassword");
	}

	// login

	public static User login(HttpServletRequest request, String name, String rawPassword) {
		var user = UserDao.login(name, rawPassword);
		if (user == null) {
			set(request, null, null);
		} else {
			set(request, user, rawPassword);
		}
		return user;
	}

	public static User reLogin(HttpServletRequest request) {
		var user = getUser(request);
		var rawPassword = getRawPassword(request);
		if (user != null) {
			return login(request, user.getName(), rawPassword);
		}
		return null;
	}

	public static void logout(HttpServletRequest request) {
		set(request, null, null);
	}

	// store user to cookie

	private static final int COOKIE_MAX_AGE = 60 * 60 * 24 * 15;

	public static void appendCookie(HttpServletResponse response, String loginName, String loginPassword) {
		var loginNameCookie = CookieUtil.make("loginName", loginName, true, COOKIE_MAX_AGE);
		var loginPasswordCookie = CookieUtil.make("loginPassword", loginPassword, true, COOKIE_MAX_AGE);
		response.addCookie(loginNameCookie);
		response.addCookie(loginPasswordCookie);
	}

	public static void removeCookie(HttpServletResponse response) {
		var loginNameCookie = CookieUtil.make("loginName", "", true, 0);
		var loginPasswordCookie = CookieUtil.make("loginPassword", "", true, 0);
		response.addCookie(loginNameCookie);
		response.addCookie(loginPasswordCookie);
	}

	public static User loginByCookieIfExist(HttpServletRequest request) {
		var user = (User) null;
		var name = CookieUtil.get(request, "loginName");
		var password = CookieUtil.get(request, "loginPassword");
		if (name != null && name.getValue() != null &&
				password != null && password.getValue() != null) {
			user = ContextUtil.login(request, CookieUtil.decrypt(name.getValue()), CookieUtil.decrypt(password.getValue()));
		}
		return user;
	}

}

File: plldb-server/src-refactor/main/java/cn/twin_kles/plldb/util/ContextUtil.kt
package cn.twin_kles.plldb.util

import cn.twin_kles.plldb.dao.UserDao
import cn.twin_kles.plldb.domain.User
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

object ContextUtil {

    // session attribute setter and getter

    fun set(request: HttpServletRequest, user: User?, rawPassword: String?) {
        request.session.setAttribute("user", user)
        request.session.setAttribute("rawPassword", rawPassword)
    }

    fun getUser(request: HttpServletRequest): User? {
        return request.session.getAttribute("user") as User?
    }

    fun getRawPassword(request: HttpServletRequest): String? {
        return request.session.getAttribute("rawPassword") as String?
    }

    // login

	fun login(request: HttpServletRequest, name: String, rawPassword: String): User? {
        val user = UserDao.login(name, rawPassword)
        set(request, user, if (user == null) null else rawPassword)
        return user
    }

	fun reLogin(request: HttpServletRequest): User? {
        val user = getUser(request)
        val rawPassword = getRawPassword(request)
        return if (user != null) {
            login(request, user.name, rawPassword!!)
        } else null
    }

	fun logout(request: HttpServletRequest) {
        set(request, null, null)
    }

    // store user to cookie

    private const val COOKIE_MAX_AGE = 60 * 60 * 24 * 15

	fun appendCookie(response: HttpServletResponse, loginName: String, loginPassword: String) {
        val loginNameCookie = CookieUtil.make("loginName", loginName, true, COOKIE_MAX_AGE)
        val loginPasswordCookie = CookieUtil.make("loginPassword", loginPassword, true, COOKIE_MAX_AGE)
        response.addCookie(loginNameCookie)
        response.addCookie(loginPasswordCookie)
    }

	fun removeCookie(response: HttpServletResponse) {
        val loginNameCookie = CookieUtil.make("loginName", "", true, 0)
        val loginPasswordCookie = CookieUtil.make("loginPassword", "", true, 0)
        response.addCookie(loginNameCookie)
        response.addCookie(loginPasswordCookie)
    }

	fun loginByCookieIfExist(request: HttpServletRequest): User? {
        var user = null as User?
        val name = CookieUtil.get(request, "loginName")
        val password = CookieUtil.get(request, "loginPassword")
        if (name != null && name.value != null && password != null && password.value != null) {
            user = login(request, CookieUtil.decrypt(name.value), CookieUtil.decrypt(password.value))
        }
        return user
    }

}